### PR TITLE
Task 4:  Caching + filtering

### PR DESCRIPTION
**Pull request by Codeassassin-Mpendulo**
Todo in this task: 

- For your data call layer, please create a in-memory caching repository of which stores a cache of data
      - Cache is only valid for 1 minute
- Add a text input filter on the nav bar of which it would do a simple contains filtering on either the title or paragraph for the list of items
      - This filtering can either but done on the front end or update your MSRV endpoint to include query params for filtering.
- Pulling down on the screen will clear cache and re-fetch data.

### CODE REVIEW

File: app/src/main/java/com/example/cocktaildictionary/actions/HomeActivityAction.kt
Comment (simphiwe-msimango) [inline app/src/main/java/com/example/cocktaildictionary/actions/HomeActivityAction.kt:1]: Just nitty picking, use the format trick I showed you.

File: app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt [hunk 1]
```diff
@@ -27,18 +33,50 @@ import java.time.LocalDateTime
...
+      private const val DONE = 0L
+      private const val ONE_SECOND = 1000L
+      private const val COUNTDOWN_TIME = 60000L
+   }
    private var myCompositeDisposable: CompositeDisposable = CompositeDisposable()
+   private val timer : CountDownTimer
    private var cacheDatabase: CacheDatabase
    private val converter:Converters = Converters()
-
+   private lateinit var cocktailList: CocktailList
```
Comment (simphiwe-msimango) [inline app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt]: Why did you add this var back?

File: app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt [hunk 2]
```diff
@@ -47,6 +85,7 @@ class HomeActivityViewModel(application: Application): AndroidViewModel(applicat
          cacheDatabase.cacheDao().addToCache(cache)
       }
       Log.d(TAG,"Successfully added to cache")
+      timer.start()
```
Comment (simphiwe-msimango) [inline app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt:88]: What happens when you close/kill the app and reopen it, does the timer kick off again?

File: app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt [hunk 3]
```diff
@@ -1,38 +1,71 @@
...
+    private lateinit var menuItem : MenuItem
+    private lateinit var globalMenu: Menu
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
         viewModel = ViewModelProvider(this)[HomeActivityViewModel::class.java]
 
+
```
Comment (simphiwe-msimango) [inline app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt]: redundant spaces buddy

File: app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt [hunk 4]
```diff
@@ -68,6 +101,7 @@ class MainActivity : AppCompatActivity() {
         binding.progressBar.visibility = View.INVISIBLE
         binding.retry.visibility = View.INVISIBLE
         binding.cocktailRecyclerview.visibility = View.VISIBLE
+//        menuItem.isVisible = true
```
Comment (simphiwe-msimango) [inline app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt]: Not advisable to push in dead code

File: app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt (hunk 2 above)
Comment (Codeassassin-Mpendulo) [reply app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt:88]: Yes another timer kicks off whilst the initial timer continues to countdown

File: app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt (hunk 3 above)
Comment (Codeassassin-Mpendulo) [reply app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt]: I removed the redundant space

File: app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt (hunk 4 above)
Comment (Codeassassin-Mpendulo) [reply app/src/main/java/com/example/cocktaildictionary/views/MainActivity.kt]: Removed the dead code

File: app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt (hunk 1 above)
Comment (Codeassassin-Mpendulo) [reply app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt]: The reason I brought this var back is because I use it in the loadFilteredData method. Its purpose is to serve as the cocktail list that will be traversed and checked if it contains the query. I tried to use the cocktail list that is stored in the cache however since the cache is cleared after 1 min it results in an error when a search query is provided when the cache is empty. I also tried to use the cocktail list from the viewState given that the viewState type is HomeActivityStates.DataLoadSuccessful however with the way my code is structured this approach results in the filtered list being displayed even after the search has been completed and the user wishes to go to the original home screen which displays all the cocktails. I also thought of going with the approach of making a get request for the cocktails and use that cocktail list however this would result in numerous unnecessary requests. Thus I opted to make use of the variable cocktail list.

Comment (simphiwe-msimango) [reply app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt]: There should be a way to filter without using a backup property. We can explore that in a later stage. For this simple app it will suffice for now.

File: app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt (hunk 2 above)
Comment (simphiwe-msimango) [reply app/src/main/java/com/example/cocktaildictionary/viewmodel/HomeActivityViewModel.kt:88]: Sounds like I can trick the app by constantly killing it and opening it. But cool. Can also be explored further later on and making it more than a minute.

**Review by simphiwe-msimango — APPROVED** [review]
